### PR TITLE
Swap individual policy assignments to `alzDefaultPolicyAssignments` module in E2E tests

### PR DESCRIPTION
**Pull request by jfaurskov**
# Overview/Summary

Update unit test pipeline to include alzDefaultPolicyAssignments

## This PR fixes/adds/changes/removes

1. Change check for policy assignments file to alzDefaultPolicyAssignments.bicep
2. Move policy assignment step to after subscription placement to allow more time for policy replication.
3. Add script to do retry logic if policy assignment fails initially. Script will retry deployment 5 times with 30 second intervals and then error out if not succeeded. 

### Breaking Changes

N/A

## Testing Evidence
Ran script manually, with custom policies not deployed to root of ALZ management group structure. 
defaultpolassign1
After failing script waits 30 seconds and then retries.
defaultpolassign2
Separately entered the policies while script was running and it succeded. 
defaultpolassign3

## As part of this Pull Request I have

- [x] Checked for duplicate [Pull Requests](https://github.com/Azure/ALZ-Bicep/pulls)
- [x] Associated it with relevant [ADO items](https://dev.azure.com/unifiedactiontracker/Solution%20Engineering/_workitems/edit/101184)
- [x] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/Azure/ALZ-Bicep/tree/main)
- [x] Performed testing and provided evidence.
- [ ] Updated relevant and associated documentation.

### CODE REVIEW

File: tests/pipelines/bicep-build-to-validate.yml [hunk 1]
```diff
@@ -152,22 +152,22 @@ jobs:
...
+        az deployment mg create --template-file infra-as-code/bicep/modules/subscriptionPlacement/subscriptionPlacement.bicep --parameters @infra-as-code/bicep/modules/subscriptionPlacement/subscriptionPlacement.parameters.example.json parTargetManagementGroupId=$(ManagementGroupPrefix)-platform-connectivity parSubscriptionIds='[$(subscriptionId)]' --location $(Location) --management-group-id $(ManagementGroupPrefix)
 
   - task: Bash@3    
-    displayName: Az CLI Subscription Placement for PR
-    name: move_sub
-    condition: and(or(ne(variables['gitManagementOUTPUT'], ''), ne(variables['gitLoggingOUTPUT'], ''), ne(variables['gitSpokeOUTPUT'], ''), ne(variables['gitHubOUTPUT'], ''), ne(variables['gitVwanOUTPUT'], '')), ne(variables['subscriptionId'], ''))
+    displayName: Az CLI Policy Assignment DINE for PR
```
Comment (jtracey93) [inline tests/pipelines/bicep-build-to-validate.yml:164]:
```suggestion
    displayName: Az CLI alzDefaultPolicyAssignments for PR
```

File: tests/pipelines/bicep-build-to-validate.yml [hunk 2]
```diff
@@ -152,22 +152,22 @@ jobs:
...
+    name: create_policy_assignment_dine
```
Comment (jtracey93) [inline tests/pipelines/bicep-build-to-validate.yml:165]:
```suggestion
    name: alz_default_policy_assignments
```

File: tests/pipelines/bicep-build-to-validate.yml [hunk 3]
```diff
@@ -152,22 +152,22 @@ jobs:
...
+        az deployment mg create --template-file infra-as-code/bicep/modules/policy/assignments/alzDefaults/alzDefaultPolicyAssignments.bicep --parameters @infra-as-code/bicep/modules/policy/assignments/alzDefaults/alzDefaultPolicyAssignments.parameters.example.json parTopLevelManagementGroupPrefix=$(ManagementGroupPrefix) parLogAnalyticsWorkSpaceAndAutomationAccountLocation=$(Location) parLogAnalyticsWorkspaceResourceID="/subscriptions/$(subscriptionId)/resourceGroups/$(ResourceGroupName)/providers/Microsoft.OperationalInsights/workspaces/alz-log-analytics" parDdosProtectionPlanId="/subscriptions/$(subscriptionId)/resourceGroups/$(ResourceGroupName)/providers/Microsoft.Network/ddosProtectionPlans/alz-Ddos-Plan" --location $(Location) --management-group-id "$(ManagementGroupPrefix)-platform"
```
Comment (jtracey93) [inline tests/pipelines/bicep-build-to-validate.yml:170]:
```suggestion
        az deployment mg create --template-file infra-as-code/bicep/modules/policy/assignments/alzDefaults/alzDefaultPolicyAssignments.bicep --parameters @infra-as-code/bicep/modules/policy/assignments/alzDefaults/alzDefaultPolicyAssignments.parameters.example.json parTopLevelManagementGroupPrefix=$(ManagementGroupPrefix) parLogAnalyticsWorkSpaceAndAutomationAccountLocation=$(Location) parLogAnalyticsWorkspaceResourceID="/subscriptions/$(subscriptionId)/resourceGroups/$(ResourceGroupName)/providers/Microsoft.OperationalInsights/workspaces/alz-log-analytics" parDdosProtectionPlanId="/subscriptions/$(subscriptionId)/resourceGroups/$(ResourceGroupName)/providers/Microsoft.Network/ddosProtectionPlans/alz-Ddos-Plan" --location $(Location) --management-group-id "$(ManagementGroupPrefix)"
```

Comment (jtracey93) [reply tests/pipelines/bicep-build-to-validate.yml:170]: should target top level MG

**Review by jtracey93 — CHANGES_REQUESTED** [review]
A few bits, also really want to ensure we have enough delay or retry logic if the policy assignments fail due to https://github.com/Azure/Enterprise-Scale/issues/902.

Can we add some error handling if the deployment fails due to this, to wait for say 5 mins and try again, as this should give replication a chance to catch up in ARM

**Comment by jfaurskov** [comment]
Looking into the error handling bits

File: .github/scripts/Set-AlzDefaultPolicyAssignment.ps1 [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+$err.clear
+while ($i -lt 4 -and $state -eq 'fail') {
+    $ErrorActionPreference = "Stop"
+    Try {
+        New-AzManagementGroupDeployment -Managementgroupid $ManagementGroupId -Location $parLocation -TemplateFile $templateFile -TemplateParameterFile $parameterFile -parTopLevelManagementGroupPrefix $parTopLevelManagementGroupPrefix -parLogAnalyticsWorkSpaceAndAutomationAccountLocation $parLogAnalyticsWorkSpaceAndAutomationAccountLocation -parLogAnalyticsWorkspaceResourceID $parLogAnalyticsWorkspaceResourceID -parDdosProtectionPlanId $parDdosProtectionPlanId
+        $state = 'success'
+    }
+    Catch {
+        $i++
+        Write-Output "Default policy assignment failed with $error"
```
Comment (jtracey93) [inline .github/scripts/Set-AlzDefaultPolicyAssignment.ps1:24]:
```suggestion
        Write-Output "ALZ Default Policy Assignments module failed to deploy with $error"
```

File: .github/scripts/Set-AlzDefaultPolicyAssignment.ps1 [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+        Write-Output "Iteration number $i"
+        Write-Output "Will retry in 30 seconds"
+        Start-Sleep -Seconds 30
+    }
+}
+If ($state -eq 'fail') {
+    Throw "Default policy assignment failed"
```
Comment (jtracey93) [inline .github/scripts/Set-AlzDefaultPolicyAssignment.ps1:31]:
```suggestion
    Throw "ALZ Default Policy Assignments module failed to deploy after $i attempts"
```

File: tests/pipelines/bicep-build-to-validate.yml [hunk 6]
```diff
@@ -28,7 +28,7 @@ jobs:
           git_diff1=$(git diff --name-only HEAD^ HEAD infra-as-code/bicep/modules/managementGroups/managementGroups.bicep)
           git_diff2=$(git diff --name-only HEAD^ HEAD infra-as-code/bicep/modules/customRoleDefinitions/customRoleDefinitions.bicep)
           git_diff3=$(git diff --name-only HEAD^ HEAD infra-as-code/bicep/modules/policy/definitions/custom-policy-definitions.bicep)
-          git_diff4=$(git diff --name-only HEAD^ HEAD infra-as-code/bicep/modules/policy/assignments/policyAssignmentManagementGroup.bicep)
+          git_diff4=$(git diff --name-only HEAD^ HEAD infra-as-code/bicep/modules/policy/assignments/alzDefaults/alzDefaultPolicyAssignments.bicep)
```
Comment (jtracey93) [inline tests/pipelines/bicep-build-to-validate.yml:31]: Should we also check for changes to the `policyAssignmentManagementGroup.bicep` module as this is used by the `alzDefaultPolicyAssignments.bicep` module multiple times?

Also should we check for changes to policy JSON files as well to trigger this? Or shall we create a separate pipeline we can trigger for these when we want to to just test policy deployment and assignment to the MG hierarchy? Something like `/azp run policy-check`?

File: tests/pipelines/bicep-build-to-validate.yml [hunk 7]
```diff
@@ -160,14 +160,17 @@ jobs:
       script: |
         az deployment mg create --template-file infra-as-code/bicep/modules/subscriptionPlacement/subscriptionPlacement.bicep --parameters @infra-as-code/bicep/modules/subscriptionPlacement/subscriptionPlacement.parameters.example.json parTargetManagementGroupId=$(ManagementGroupPrefix)-platform-connectivity parSubscriptionIds='[$(subscriptionId)]' --location $(Location) --management-group-id $(ManagementGroupPrefix)
 
-  - task: Bash@3    
+  - task: AzurePowerShell@5
     displayName: Az CLI Policy Assignment DINE for PR
```
Comment (jtracey93) [inline tests/pipelines/bicep-build-to-validate.yml:164]:
```suggestion
    displayName: Az CLI alzDefaultPolicyAssignments for PR
```

**Review by jtracey93 — CHANGES_REQUESTED** [review]
Just a few bits 👍

Great work though 👍

File: tests/pipelines/bicep-build-to-validate.yml (hunk 6 above)
Comment (jfaurskov) [reply tests/pipelines/bicep-build-to-validate.yml:31]: +1 on the policyAssignmentManagementGroup.bicep. For the individual policy files, I think that one loops into another conversation around policies that we are currently having and so would be outside the scope of this pipeline.

**Pull request merged**